src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/CrossTarget/TargetFrameworkProvider.cs

```diff
@@ -37,20 +38,17 @@ public ITargetFramework GetTargetFramework(string shortOrFullName)
             {
                 lock (_targetsLock)
                 {
-                    // use linear search here, since there not many target frameworks and it would most efficient.
-                    targetFramework = _cachedTargetFrameworks.FirstOrDefault(x => x.Equals(shortOrFullName));
-                    if (targetFramework != null)
+                    if (!TryGetCachedTargetFramework(shortOrFullName, out targetFramework))
                     {
-                        return targetFramework;
-                    }
-
-                    var frameworkName = _nuGetFrameworkParser.ParseFrameworkName(shortOrFullName);
-                    if (frameworkName != null)
-                    {
-                        var shortName = _nuGetFrameworkParser.GetShortFrameworkName(frameworkName);
-                        targetFramework = new TargetFramework(frameworkName, shortName);
-                        // remember target framework - there can not bee too many of them across the solution.
-                        _cachedTargetFrameworks.Add(targetFramework);
+                        var frameworkName = _nuGetFrameworkParser.ParseFrameworkName(shortOrFullName);
+                        if (frameworkName != null && 
+                            !TryGetCachedTargetFramework(frameworkName.FullName, out targetFramework))
```

Comment:
>TryGetCachedTargetFramework [](start = 29, length = 27)

FYI: This change is not really what i meant here when created this component. I was going to have a cache of unique TFMs, when now you could have same TFMs but specified by different short/long names or their aliases. Instead of doing this short term fix i would do a better TFM recognition in the line 43: var frameworkName = _nuGetFrameworkParser.ParseFrameworkName(shortOrFullName);
i.e. improve the way we parse and normalize to some common standard framework name (potentially keeping actual "alias" provided in the project).

Reply:
I agree we should be keeping the original alias. I'd prefer to call that the `ShortName`, and have a separate `CanonicalShortName` for the value returned by `_nuGetFrameworkParser.GetShortFrameworkName`, but I would need to ensure this does not affect semantics elsewhere.

Reply:
Why should we be keeping the original alias? We'll only be tempted to compare to it. I've got a change where I throwaway usage of shortname completely apart from display purposes. The FullName is the canonical name.

Reply:
The primary purpose of keeping the alias here would be to avoid one extra check -- since the non-canonical shortnames do not occur in the cache, you have to parse the full name every time to get a cache hit. I don't think this is a bottleneck though so I'm fine leaving out this alias, and relying on the full name. This is ready to merge if I can get another signoff.

Reply:
There's three different names here:

- User written short name
- Canonical Full name
- Canonical short name

Only the former two should used, the latter _should_ never be used.

**User written short name (`<TargetFramework>` property)**: Used in UI for display purposes and passed to NuGet so that it can write conditions in nuget.g.props/targets. It should never ever be interpreted, parsed or exchanged with any other feature.

**Canonical full name (`<TargetFrameworkMoniker>` property):**: The canonical name that is used everywhere else and should be the exchange between Visual Studio features.

**Canonical short name:** No MSBuild representation. This a concept that only exists in NuGet, and only used within NuGet packages. It should not be used anywhere else - the logic for interpreting/producing it only exists in NuGet.

Reply:
The issue is that this method is called with a "shortOrFullName" so we don't necessarily have the "user written short name" (`<TargetFramework>` property). I'll hold this change till I get back from vacation.